doc/mgmt/SONiC_PIM_SM_HLD.md

```diff
@@ -0,0 +1,1036 @@
+
+### 3.1.2 ipmcorchagent changes
+As part of the PIM ASM support following functionalities attained by ipmcorchagent,<br> 
+1. CPU bit update for (*,G) and (S,G) routes<br> 
+2. Updating the KAT timer value for (S,G) routes from FRR to ASIC<br> 
+
+#### 3.1.2.1 CPU bit update
+
+CPU bit must be updated for (*,G) or (S,G) entry to punt the packets to FRR stack using CPU port. <br>
+To achieve this functionality during creation of IPMC entry in BCM we need to set the action as SAI_PACKET_ACTION_COPY  .<br>
+If CPU port removed from the (*,G) or (S,G) entry then ipmcorchagent need to set action to SAI_PACKET_ACTION_COPY_CANCEL <br>
+
+Following are the sequence of events handled in ipmcorchagent for CPU bit update,<br>
+1. ipmcfpmsyncd update APPL_DB with (S,G) or (*,G) and OIF as "pimreg" in APP_IPMC_ROUTE_TABLE_NAME table
+2. ipmcorchagent gets notified with APP_IPMC_ROUTE_TABLE_NAME<br>
+3. If operation is SET with ifname as "pimreg", set the ACTION as SAI_PACKET_ACTION_COPY
+4. If operation is DELETE with ifname as "pimreg", set the ACTIOn as SAI_PACKET_ACTION_COPY_CANCEL<br>
+
+| Existing Entry  |   New Entry | Operation | Action |
+| ----------------|-------------|-----------|--------|
+| None            | (*,G) or (S,G) OIF as pimreg | SET | SAI_PACKET_ACTION_COPY (IPMC entry created) |
+| (*,G) or (S,G) OIF as pimreg | (*,G) or (S,G) OIF as pimreg | DEL | IPMC entry removed |
+| None            | (*,G) or (S,G) OIF as Vlan10 | SET | SAI_PACKET_ACTION_FORWARD (IPMC entry created) |
+| (*,G) or (S,G) OIF as Vlan10 | (*,G) or (S,G) OIF as Vlan10,pimreg | SET | SAI_PACKET_ACTION_COPY |
+| (*,G) or (S,G) OIF as Vlan10,pimreg | (*,G) or (S,G) OIF as Vlan10 | SET | SAI_PACKET_ACTION_COPY_CANCEL |
+| (*,G) or (S,G) OIF as Vlan10,pimreg | (*,G) or (S,G) OIF as Vlan10,pimreg | DEL | IPMC entry removed |
+| (*,G) or (S,G) OIF as pimreg | (*,G) or (S,G) OIF as pimreg | DEL | IPMC entry removed |
+
+#### 3.1.2.2 IIF expansion in OIF List
+
+IIF member ports to be included in IPMC group member on following cases,
+1. L3 multicast case, local receiver on the IIF vlan then add IIF vlan in OIF list
+2. Shared case (L3 and Snooping enabled), local receiver on the IIF vlan then add IIF vlan member ports into OIF list
```

Comment:
With respect to L3 multicast and Snooping enabled case, more explanation might be needed as both can have same entry and what happens if snooping on specific VLAN enables/disables and without snooping enabled and just with PIM. if its captured in some other document , please share it.

Reply:
In Orchagent, there are two cases handled (1 Only PIM enabled, 2 Both PIM and Snooping enabled).... Both case for if IIF is VLAN, VLAN is updated in OIF list.
ipmcorchagent, upon receiving the update adds IIF VLAN to OIF list(class MRouteEntry and member list m_oifs updated with IIF vlan)
a) calls IPMC group creation incase of L3 VLAN
b) call L2MC group creation incase of snooping enabled on the VLAN